portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that handles nicknaming. */
+@WebServlet("/nickname")
+public class NicknameServlet extends HttpServlet {
+
+  /** Returns the current user's nickname. */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/index.html");
+      return;
+    }
+    String id = userService.getCurrentUser().getUserId();
+    response.getWriter().println(getNickname(id));
+  }
+
+  /** Updates the current user's nickname. */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/index.html");
+      return;
+    }
+
+    String nickname = request.getParameter("nickname");
+    String id = userService.getCurrentUser().getUserId();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity entity = new Entity("UserInfo", id);
```

Comment:
Just checking, the POST to `/nickname` should be done after `/add-user` right?  Does it still make sense to create a new UserInfo Entity here, or should it be looked up from the datastore?  Also if an identity with the same id already works will this .put() call update the existing identity, or create a new one, or result in an error?

Reply:
Yes, `/nickname` gets requested after `/add-user`.

The example from the STEP codebase commented that `The put() function automatically inserts new data or updates existing data based on ID`. I probably would have done a Query as per your suggestion if I hadn't seen this beforehand. So, for the last question, this should update the existing identity.

In my opinion, I like this version, because in either case, I would have to use `.put()`, but this doesn't require a Query. Let me know if you think otherwise or have other suggestions!

Reply:
OK, this sounds good.  Maybe just add a comment that .put() will either update or create a new datastore entry

Reply:
Will do.